File: from-0601-to-0700/0650-2-keys-keyboard/solution.go
func minSteps(n int) int {
    dp := make([]int, n + 1)
    for i := 0; i <= n; i ++ {
        dp[i] = 1000
    }
    dp[1] = 0

    for i := 2; i < n + 1; i ++ {
        for j := 1; j < i / 2 + 1; j ++ {
            if i % j == 0 {
                dp[i] = min(dp[i], dp[j] + i / j)
            }
        }
    }
    return dp[n]
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: from-0601-to-0700/0650-2-keys-keyboard/solution.py
class Solution:
    def minSteps(self, n: int) -> int:
        if n == 1:
            return 0
        cache = dict()
        def helper(count, paste):
            if count == n:
                return 0
            if count > n:
                return 1000
            if (count, paste) in cache:
                return cache[(count, paste)]

            # Paste
            resPaste = 1 + helper(count + paste, paste)
            # Copy & Paste
            resCopyPaste = 2 + helper(count + count, count)

            cache[(count, paste)] = min(resPaste, resCopyPaste)
            return cache[(count, paste)]
            
        return 1 + helper(1, 1)